CommonTools/RecoAlgos/src/PrimaryVertexAssignment.cc

```diff
@@ -89,13 +107,26 @@ std::pair<int, PrimaryVertexAssignment::Quality> PrimaryVertexAssignment::charge
   // first use "closest in Z" with tight cuts (targetting primary particles)
   const float add_cov = vtxIdMinSignif >= 0 ? vertices[vtxIdMinSignif].covariance(2, 2) : 0.f;
   const float dzE = sqrt(track->dzError() * track->dzError() + add_cov);
-  if (vtxIdMinSignif >= 0 and
+  if (!fOnlyUseFirstDz_ and vtxIdMinSignif >= 0 and
       (dzmin < maxDzForPrimaryAssignment_ and dzmin / dzE < maxDzSigForPrimaryAssignment_ and
        track->dzError() < maxDzErrorForPrimaryAssignment_) and
       (!useTime or dtmin / timeReso < maxDtSigForPrimaryAssignment_)) {
     iVertex = vtxIdMinSignif;
   }
 
+  // consider only distances to first vertex for association of pileup vertices (originally used in PUPPI)
+  if ((fOnlyUseFirstDz_) && (vtxIdMinSignif >= 0) && (std::abs(track->eta()) > fEtaMinUseDz_))
+    iVertex = ((std::abs(track->dz(vertices.at(0).position())) < maxDzForPrimaryAssignment_ and
+                std::abs(track->dz(vertices.at(0).position())) / dzE < maxDzSigForPrimaryAssignment_ and
+                track->dzError() < maxDzErrorForPrimaryAssignment_) and
+               (!useTime or std::abs(time - vertices.at(0).t()) / timeReso < maxDtSigForPrimaryAssignment_))
+                  ? 0
+                  : vtxIdMinSignif;
+
+  // protect high pT particles from association to pileup vertices and assign them to the first vertex
+  if ((fPtMaxCharged_ > 0) && (vtxIdMinSignif >= 0) && (track->pt() > fPtMaxCharged_))
```

Comment:
This could be evaluated first, so that the previous computations can be avoided

Reply:
I see that below (in the part of the code not modified by you) the more expensive `deltaR` is used instead of `deltaR2`: you can profit and also apply this simple fix...